File: OOP/JAVa/Matrix2dArray.java

public class Matrix2dArray implements Matrix {

    private double[][] data;

    public int getHeight(){
        return data.length;
    }

    public int getWidth(){
        return data[0].length;
    }

    public Matrix2dArray(int height, int width){
        data = new double[height][width];
    }
    
    public double getElem(int i, int j){
        return data[i][j];
    }

    public void setElem(int i, int j, double value){
        data[i][j] = value;
    }

    public Matrix add(Matrix other){
        int rows = data.length;
        int columns = data[0].length;
        Matrix2dArray result = new Matrix2dArray(rows, columns);
        for (int i = 0; i < rows; i++){
            for (int j = 0; j < columns; j++){
                result.data[i][j] = data[i][j] + other.getElem(i, j);
            }
        }

        return result;
    }

    public Matrix multiply(Matrix other){
        int rows = data.length;
        int columns = data[0].length;
        Matrix2dArray result = new Matrix2dArray(rows, other.getWidth());

        for (int i = 0; i < rows; i++){
            for (int j = 0; j < other.getWidth(); j++){
                result.data[i][j] = 0;
                for (int k = 0; k < columns; k++ ){
                    result.data[i][j] += data[i][k] * other.getElem(k, j);
                }
            }
        }

        return result;

    }

    @Override
    public String toString(){
        String output = "";
        for (int i = 0; i < data.length; i++){
            for (int j = 0; j < data[0].length; j++){
                output += data[i][j] + "\t";
            }
            output += "\n";
        }

        return output;
    }

    
}

File: OOP/JAVa/Matrix2DArray.cpp
#include"bits\stdc++.h"
#include "Matrix.cpp"
class Matrix2DArray :public Matrix
{   int height1;
    int width1;
    double **data;
    public:
     Matrix2DArray(){}
     Matrix2DArray(int height,int width)
     {  height1=height;
        width1=width;
        data=new double *[height];
        for (int i=0;i<width;i++)
        {
           data[i]=new double [width];
        }
     }
     int min(int a,int b)
     {
        if (data[a]<data[b]) return a;
        return b;
     }
     double getElement(int i,int j)
     {
        return data[i][j];
     }
     void setElement(int i,int j,double value)
     {
        data[i][j]=value;
     }
Matrix * add (Matrix *other )
{ Matrix2DArray * result=new  Matrix2DArray(height1,width1);
  for (int i=0;i<height1;i++)
  {
     for (int j=0;j<width1;j++)
     {  
        result->data[i][j]=data[i][j]+other->getElement(i,j);
     }
  }
  return result;
}

Matrix * multiply(Matrix * other)
{
      Matrix2DArray * Result1=new Matrix2DArray(height1,width1);
      for (int i=0;i<height1;i++)
      {
         for(int j=0;j< width1;j++)
         {
            Result1->data[i][j]=0;
         
         for (int k=0;k<width1;k++) 
         {
            Result1->data[i][j]+=data[i][k] * other->getElement(k,i);
         }

         }

      }
    return Result1;
}
friend ostream & operator  <<  (ostream & out ,const Matrix2DArray & M2D)  
{
   for (int i=0;i<M2D.height1;i++)
   {  for (int j=0;j<M2D.width1;j++)
      { out<<M2D.data[i][j]<<"     ";
      }
      cout<<endl;
   }
   return out;
}

   
};